### ISSUE TITLE
Review plugin API and publish as NuGet package

### ISSUE THREAD

**Issue by mast-eu** [gitextensions/gitextensions#6624]
Extracted from https://github.com/gitextensions/gitextensions/issues/6542#issuecomment-493647677 (by @RussKie ):

I think before we can start moving any plugins we need to define a public plugin contract. This would involve moving and publishing `GitUIPluginInterfaces` project as a package.
To do that, we need to review `IGitModule` and all other interfaces - should they be there, do they have all necessary definitions and so on.
I would also like to rename `GitUIPluginInterfaces` to something like `GitExtensions.Plugins.Interfaces` (open to ideas).

**Comment by apuch** [comment]
I really would like a stable plugin API. I maintain a small plugin tailored for our inhouse usage. I just need to call a few git commands. With every release, I have to fiddle out how it is done now (doing this since 2.48). Dont get me wrong, I am thankful that the possibility exists.

**Comment by mast-eu** [comment]
@apuch: you may already have noticed that we are doing a major overhaul to plugins. See #6542 for more details, if you are interested.

> With every release, I have to fiddle out how it is done now

May I ask what methods you are using for your in-house plugin that are currently not stable?

> Dont get me wrong, I am thankful that the possibility exists.

Don't worry. Your observation is correct and we are working towards improving the situation for people like you (and ourselves).

**Comment by apuch** [comment]
> May I ask what methods you are using for your in-house plugin that are currently not stable?

The breaking change in 3.0 (of course) and  `m_currentGitUiCommands.GitModule.GitExecutable.GetOutput(cmd)`

Ideally, I could get the output from the git process when it happens to get some kind of progress indicator and a way to cancel a running process. 

With 2.X, I got somehow a process handle (cant remember exactly) and could achieve both with that but this probably exposes to much of the internals. With 3.x, there is most likely also a way to achieve this but tbh I haven't digged too deep into the API, just copy&paste from an official plugin.

**Comment by RussKie** [comment]
@apuch you feedback is valuable (I don't really use any plugins). 
Feel free to get involved, the more - the merrier.

**Comment by Cussa** [comment]
Hello people. I started to use the GitExtensions last month, and I like it. I had some ideas about creating some plugins to automate some processes that I have. But I found no documentation about how to do that.
When I tried to find something on the repo, I found this here.
Is there any start point that you could give to me?

**Comment by pmiossec** [comment]
@Cussa I'm not the better person to answer but as a quick answer, you clone and use  this project to develop your own plugin: 

https://github.com/gitextensions/gitextensions.plugintemplate

**Comment by RussKie** [comment]
@Cussa, awesome! :)

There are few pieces you'd want to have a look at:
- a Plugin Template is located at https://github.com/gitextensions/gitextensions.plugintemplate (thanks @pmiossec)
- new plugins are installed via the Plugin Manager (https://github.com/gitextensions/gitextensions.pluginmanager), which is shipped as part of 3.2.1 bundle.

The Template repo has some very basic docs, but by far and large we have a gap, and if you could capture your experience and contribute it back - it would be highly appreciated.

@maraf and @mast-eu are our main extensibility champions, and they have developed few plugins. If you get stuck and have questions hola on our gitter channel, and we'll do our best to help.

**Comment by mast-eu** [comment]
@Cussa Welcome :smiley: 

You might start by generating a new repo from our template repo. This can be done from here: https://github.com/gitextensions/gitextensions.plugintemplate/generate

Afterwards, open the solution in VS, compile it and run it. 
This will open a new instance of Git Extensions. Open the `Plugins` menu and inside you will find an `Empty Plugin`. It only shows a MessageBox when you click on it. Nothing more. But it's a good starting point to see how things work.
From there on you can start modifying it
The VS solution as you like.

Don't hesitate to ask for help if you get stuck.

**Comment by Cussa** [comment]
Hello @mast-eu @RussKie ! I already started to take a look at them. I found a small "problem" because I was trying to create the plugin and use the Plugin Manager with a local folder as a nuget repo, which didn't work.

On my creating a plugin similar to the Release Notes Generator, but the idea is to go forward with it and have other features on it, like the MarkDown, the possibility to understand the commit types and generate a better Release notes.

I'm working now with the internal logic of the library to do it and I believe that during the week I'll start to touch the plugin itself. I'll take my notes to try to improve the documentation if it's possible. And I'll give later my impressions about it.

Thanks in advance for the so good support.

**Comment by mast-eu** [comment]
> I found a small "problem" because I was trying to create the plugin and use the Plugin Manager with a local folder as a nuget repo, which didn't work.

This should actually work. Please open a new issue in https://github.com/gitextensions/gitextensions.pluginmanager if it doesn't and we'll figure it out.

> On my creating a plugin similar to the Release Notes Generator, but the idea is to go forward with it and have other features on it, like the MarkDown, the possibility to understand the commit types and generate a better Release notes.

Sounds interesting. Looking forward to see it.

> I'll take my notes to try to improve the documentation if it's possible. And I'll give later my impressions about it.

:+1: This will be highly appreciated.

**Comment by Cussa** [comment]
So, I'm working on the plugin, and I have already some small considerations.

- Most of the plugins will have some UI. So, it makes sense to have the template with some UI already.
  - In my case, I cloned the Plugin Manager, and I'm using that as a reference to create mine.
- Most of the plugins will have some Settings. So, it makes sense to have the template with some Settings examples.
   - Again, I'm using the Plugin Manager as a model. However, as it has only one type of setting, It would be handy to have some examples.
- It would be good to include some Script to rename the points to the plugin name.
   - When you use the template from GitHub, it generates the repo, but you have to go everywhere to rename files and folders to your plugin name. I'm thinking about some simple script that could do that easily.

**Comment by mast-eu** [comment]
Thanks for the (first round of) feedback. Much appreciated. I've created a new issue from it: https://github.com/gitextensions/gitextensions.plugintemplate/issues/11

Let's continue discussing the template-specific aspects over there, because otherwise this issue gets too far off topic.
I'll answer you soon in the new issue.
